File: src/ai/runtimeContext.ts
/**
 * Runtime / user-preference context injected into AI system prompts.
 * Secrets (API keys, MCP tokens) are never included.
 */
import type { AppSettings } from "../domain/types";
import { client } from "../infra/client";
import { getAppVersionCached, getSettingsCached } from "../infra/settingsCache";
import { ensureDetectedTimezone } from "../infra/timezoneCache";

export interface AiRuntimeContextDto {
  app_name: string;
  app_version: string;
  os_family: string;
  os_name: string;
  os_version: string;
  arch: string;
  locale: string;
  theme: string;
  timezone_setting: string;
  timezone_resolved: string;
  sound_enabled: boolean;
  notify_on_failure: boolean;
  launch_minimized: boolean;
  auto_backup_on_start: boolean;
  log_retention_days: number;
  ai_provider: string;
  ai_model: string;
  ai_base_host: string;
  mcp_enabled: boolean;
  mcp_listen: string;
  now_local: string;
  now_utc: string;
  shell_hint: string;
  path_sep: string;
  config_dir: string;
  data_dir: string;
  notes: string[];
}

let cache: { at: number; block: string } | null = null;
const TTL_MS = 30_000;

export function formatRuntimeContextBlock(ctx: AiRuntimeContextDto): string {
  const notes = (ctx.notes ?? []).map((n) => `- ${n}`).join("\n");
  return `<callai_runtime_context>
app: ${ctx.app_name} v${ctx.app_version}
os: ${ctx.os_family} / ${ctx.os_name} ${ctx.os_version} (${ctx.arch})
locale: ${ctx.locale}
theme: ${ctx.theme}
timezone.setting: ${ctx.timezone_setting}
timezone.resolved: ${ctx.timezone_resolved}
now.local: ${ctx.now_local}
now.utc: ${ctx.now_utc}
shell: ${ctx.shell_hint}
path_sep: ${ctx.path_sep}
dirs.config: ${ctx.config_dir}
dirs.data: ${ctx.data_dir}
prefs.sound_enabled: ${ctx.sound_enabled}
prefs.notify_on_failure: ${ctx.notify_on_failure}
prefs.launch_minimized: ${ctx.launch_minimized}
prefs.auto_backup_on_start: ${ctx.auto_backup_on_start}
prefs.log_retention_days: ${ctx.log_retention_days}
ai.provider: ${ctx.ai_provider}
ai.model: ${ctx.ai_model}
ai.base_host: ${ctx.ai_base_host}
mcp.enabled: ${ctx.mcp_enabled}
mcp.listen: ${ctx.mcp_listen}
notes:
${notes}
</callai_runtime_context>

Treat this block as authoritative environment/preferences for generation.
Pick binaries, paths, schedule times, and UI copy that fit this machine and locale.
Do not invent paths outside dirs.config / dirs.data unless the user asked.
Never echo or request secrets.`;
}

/** Browser/mock fallback when Tauri command is unavailable. */
export async function buildBrowserRuntimeContext(): Promise<AiRuntimeContextDto> {
  const settings: AppSettings = await getSettingsCached().catch(
    () =>
      ({
        theme: "system",
        locale: "zh-CN",
        launch_minimized: false,
        log_retention_days: 30,
        notify_on_failure: false,
        sound_enabled: true,
        timezone: "system",
        auto_backup_on_start: true,
        backup_keep_count: 10,
        ai: {
          provider: "openai",
          base_url: "https://api.openai.com/v1",
          api_key: "",
          model: "gpt-5.6-terra",
        },
        mcp: {
          enabled: false,
          listen_host: "127.0.0.1",
          port: 33927,
          auth_token: "",
        },
      }) as AppSettings,
  );
  const version = await getAppVersionCached().catch(() => "0.0.0");
  const tzResolved = await ensureDetectedTimezone().catch(() => "UTC");
  const ua = typeof navigator !== "undefined" ? navigator.userAgent : "";
  let os_family = "unknown";
  if (/Mac/i.test(ua)) os_family = "macos";
  else if (/Win/i.test(ua)) os_family = "windows";
  else if (/Linux/i.test(ua)) os_family = "linux";

  const base = settings.ai?.base_url ?? "";
  const host =
    base.replace(/^https?:\/\//, "").split("/")[0] || "(unset)";

  const now = new Date();
  return {
    app_name: "callai",
    app_version: version,
    os_family,
    os_name: os_family,
    os_version: "browser",
    arch: typeof navigator !== "undefined" ? navigator.platform : "unknown",
    locale: settings.locale || "zh-CN",
    theme: settings.theme || "system",
    timezone_setting: settings.timezone || "system",
    timezone_resolved: tzResolved,
    sound_enabled: settings.sound_enabled !== false,
    notify_on_failure: !!settings.notify_on_failure,
    launch_minimized: !!settings.launch_minimized,
    auto_backup_on_start: settings.auto_backup_on_start !== false,
    log_retention_days: settings.log_retention_days ?? 30,
    ai_provider: settings.ai?.provider ?? "openai",
    ai_model: settings.ai?.model ?? "",
    ai_base_host: host,
    mcp_enabled: !!settings.mcp?.enabled,
    mcp_listen: `${settings.mcp?.listen_host ?? "127.0.0.1"}:${settings.mcp?.port ?? 33927}`,
    now_local: now.toString(),
    now_utc: now.toISOString(),
    shell_hint:
      os_family === "windows"
        ? "powershell / cmd"
        : os_family === "macos"
          ? "zsh/bash; open, osascript, say"
          : "bash/sh",
    path_sep: os_family === "windows" ? "\\" : "/",
    config_dir: "(browser mock)",
    data_dir: "(browser mock)",
    notes: [
      "Never echo or request API keys / MCP tokens.",
      "For pure chime reminders use binary `__callai_alarm__`.",
      "Respond in the user's locale when writing names/copy.",
    ],
  };
}

export async function loadRuntimeContextBlock(force = false): Promise<string> {
  if (!force && cache && Date.now() - cache.at < TTL_MS) {
    return cache.block;
  }
  try {
    const ctx = await client.getAiRuntimeContext();
    const block = formatRuntimeContextBlock(ctx);
    cache = { at: Date.now(), block };
    return block;
  } catch {
    const block = formatRuntimeContextBlock(await buildBrowserRuntimeContext());
    cache = { at: Date.now(), block };
    return block;
  }
}

export function clearRuntimeContextCache(): void {
  cache = null;
}

File: src-tauri/src/domain/runtime_context.rs
//! Runtime snapshot for AI prompt injection (OS, locale, prefs — no secrets).

use serde::{Deserialize, Serialize};

use super::{AppSettings, LocaleCode, ThemeMode};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct AiRuntimeContext {
    pub app_name: String,
    pub app_version: String,
    pub os_family: String,
    pub os_name: String,
    pub os_version: String,
    pub arch: String,
    pub locale: String,
    pub theme: String,
    pub timezone_setting: String,
    pub timezone_resolved: String,
    pub sound_enabled: bool,
    pub notify_on_failure: bool,
    pub launch_minimized: bool,
    pub auto_backup_on_start: bool,
    pub log_retention_days: u32,
    pub ai_provider: String,
    pub ai_model: String,
    /// Base URL host only (never the API key).
    pub ai_base_host: String,
    pub mcp_enabled: bool,
    pub mcp_listen: String,
    pub now_local: String,
    pub now_utc: String,
    pub shell_hint: String,
    pub path_sep: String,
    pub config_dir: String,
    pub data_dir: String,
    pub notes: Vec<String>,
}

impl AiRuntimeContext {
    pub fn collect(settings: &AppSettings, config_dir: String, data_dir: String) -> Self {
        let info = os_info::get();
        let os_family = if cfg!(target_os = "macos") {
            "macos"
        } else if cfg!(target_os = "windows") {
            "windows"
        } else if cfg!(target_os = "linux") {
            "linux"
        } else {
            std::env::consts::OS
        }
        .to_string();

        let os_name = info.os_type().to_string();
        let os_version = info.version().to_string();
        let arch = std::env::consts::ARCH.to_string();

        let tz_setting = settings.timezone().to_string();
        let tz_resolved = crate::domain::resolve_timezone(settings.timezone())
            .map(|tz| tz.name().to_string())
            .unwrap_or_else(|_| crate::domain::detect_system_timezone().name().to_string());

        let now_utc = chrono::Utc::now();
        let now_local = crate::domain::resolve_timezone(settings.timezone())
            .ok()
            .map(|tz| now_utc.with_timezone(&tz).to_rfc3339())
            .unwrap_or_else(|| now_utc.to_rfc3339());

        let shell_hint = match os_family.as_str() {
            "windows" => "powershell / cmd; prefer PowerShell for scripts".into(),
            "macos" => "zsh/bash; open, osascript, say available".into(),
            "linux" => "bash/sh; notify-send may be available".into(),
            _ => "posix shell".into(),
        };

        let path_sep = if cfg!(windows) { "\\" } else { "/" }.to_string();

        let locale = match settings.locale() {
            LocaleCode::ZhCn => "zh-CN",
            LocaleCode::En => "en",
        }
        .to_string();

        let theme = match settings.theme() {
            ThemeMode::System => "system",
            ThemeMode::Light => "light",
            ThemeMode::Dark => "dark",
        }
        .to_string();

        let base = settings.ai.base_url.trim();
        let ai_base_host = url_host(base).unwrap_or_else(|| base.chars().take(64).collect());

        let mut notes = vec![
            "Never echo or request API keys / MCP tokens.".into(),
            "For pure chime reminders use binary `__callai_alarm__`.".into(),
            "Respond in the user's locale when writing names/copy.".into(),
        ];
        if os_family == "macos" {
            notes.push("macOS: prefer `say`, `osascript`, `open`; avoid Windows-only cmds.".into());
        } else if os_family == "windows" {
            notes.push("Windows: prefer `powershell` / `pwsh`; avoid bash-only one-liners.".into());
        } else if os_family == "linux" {
            notes.push("Linux: prefer portable shell; check binary exists before using.".into());
        }

        Self {
            app_name: "callai".into(),
            app_version: env!("CARGO_PKG_VERSION").into(),
            os_family,
            os_name,
            os_version,
            arch,
            locale,
            theme,
            timezone_setting: tz_setting,
            timezone_resolved: tz_resolved,
            sound_enabled: settings.notify.sound_enabled,
            notify_on_failure: settings.notify.notify_on_failure,
            launch_minimized: settings.runtime.launch_minimized,
            auto_backup_on_start: settings.backup.auto_backup_on_start,
            log_retention_days: settings.runtime.log_retention_days,
            ai_provider: settings.ai.provider.as_str().to_string(),
            ai_model: settings.ai.model.clone(),
            ai_base_host,
            mcp_enabled: settings.mcp.enabled,
            mcp_listen: settings.mcp.listen_addr(),
            now_local,
            now_utc: now_utc.to_rfc3339(),
            shell_hint,
            path_sep,
            config_dir,
            data_dir,
            notes,
        }
    }

    /// Markdown-ish block injected into AI system prompts.
    pub fn to_prompt_block(&self) -> String {
        let notes = self
            .notes
            .iter()
            .map(|n| format!("- {n}"))
            .collect::<Vec<_>>()
            .join("\n");
        format!(
            r#"<callai_runtime_context>
app: {app} v{ver}
os: {family} / {name} {os_ver} ({arch})
locale: {locale}
theme: {theme}
timezone.setting: {tz_set}
timezone.resolved: {tz_res}
now.local: {now_local}
now.utc: {now_utc}
shell: {shell}
path_sep: {sep}
dirs.config: {cfg}
dirs.data: {data}
prefs.sound_enabled: {sound}
prefs.notify_on_failure: {notify}
prefs.launch_minimized: {min}
prefs.auto_backup_on_start: {backup}
prefs.log_retention_days: {ret}
ai.provider: {ai_p}
ai.model: {ai_m}
ai.base_host: {ai_h}
mcp.enabled: {mcp_on}
mcp.listen: {mcp_listen}
notes:
{notes}
</callai_runtime_context>

Treat this block as authoritative environment/preferences for generation.
Pick binaries, paths, schedule times, and UI copy that fit this machine and locale.
Do not invent paths outside dirs.config / dirs.data unless the user asked.
"#,
            app = self.app_name,
            ver = self.app_version,
            family = self.os_family,
            name = self.os_name,
            os_ver = self.os_version,
            arch = self.arch,
            locale = self.locale,
            theme = self.theme,
            tz_set = self.timezone_setting,
            tz_res = self.timezone_resolved,
            now_local = self.now_local,
            now_utc = self.now_utc,
            shell = self.shell_hint,
            sep = self.path_sep,
            cfg = self.config_dir,
            data = self.data_dir,
            sound = self.sound_enabled,
            notify = self.notify_on_failure,
            min = self.launch_minimized,
            backup = self.auto_backup_on_start,
            ret = self.log_retention_days,
            ai_p = self.ai_provider,
            ai_m = self.ai_model,
            ai_h = self.ai_base_host,
            mcp_on = self.mcp_enabled,
            mcp_listen = self.mcp_listen,
            notes = notes,
        )
    }
}

fn url_host(base: &str) -> Option<String> {
    let b = base.trim().trim_end_matches('/');
    if b.is_empty() {
        return None;
    }
    // crude parse without full URL crate dependency
    let rest = b
        .strip_prefix("https://")
        .or_else(|| b.strip_prefix("http://"))
        .unwrap_or(b);
    let host = rest.split('/').next().unwrap_or(rest);
    if host.is_empty() {
        None
    } else {
        Some(host.to_string())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::domain::AppSettings;

    #[test]
    fn prompt_block_contains_markers() {
        let ctx = AiRuntimeContext::collect(&AppSettings::default(), "/cfg".into(), "/data".into());
        let block = ctx.to_prompt_block();
        assert!(block.contains("<callai_runtime_context>"));
        assert!(block.contains("app: callai"));
        assert!(block.contains("timezone.resolved:"));
        assert!(!block.contains("api_key"));
    }

    #[test]
    fn url_host_strips_path() {
        assert_eq!(
            url_host("https://api.openai.com/v1"),
            Some("api.openai.com".into())
        );
    }
}
